### PR TITLE
Infinite scroll for AMP

### PR DESCRIPTION
**Pull request by pereirinha**
This PR adds the foundations for AMP compatibility for Infinite Scroll and support for `twentynineteen` and `twentytwenty` themes.

#### Changes proposed in this Pull Request:

* This PR adds support for Infinite Scroll on AMP via `amp-next-page`;
* This PR adds support for `twentynineteen` and `twentytwenty` themes;

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This adds support for Infinite Scroll on AMP requests;

#### Testing instructions:

1. Ensure the [AMP](https://wordpress.org/plugins/amp/) plugin is installed and enabled;
1. On Jetpack Settings > Writing  > Theme Enhancements > Infinite Scroll, set the radio to either `Load more posts in page with a button` or `Load more posts as the reader scrolls down`. Due to [AMP implementation](https://amp.dev/documentation/components/amp-next-page/#recommendation-box-(more-suggested-links)) of the functionality, the next page will always be autoloaded and still show the Older Posts button, if the user scrolls faster then the next page loads;
1. Enable `twentynineteen` or `twentytwenty` theme;
1. Make sure you have enough posts to enable pagination;
1. Visit the home page or an archive page;
1. Scroll to the footer. The next page should be loaded by then, or you can click the older posts button;

#### Proposed changelog entry for your changes:

* Adds support for Infinite Scroll on `twentynineteen` and `twentytwenty` themes for AMP requests;
* Add the following filters to allow extending support to other themes:
    * `jetpack_amp_infinite_footers`
get the theme footers to be added to `amp-next-page` footer component;
    * `jetpack_amp_infinite_output`
filters the output buffer to allow inserting `next-page-hide` attributes and remove the filtered footers from the standard output;
    * `jetpack_amp_infinite_separator`
filters the `amp-next-page` separator markup;
    * `jetpack_amp_infinite_older_posts`
filters the `amp-next-page` older posts button markup;
    * `jetpack_amp_infinite_next_page_data`
filters the `amp-next-page` data (title, url and image);

### CODE REVIEW

File: modules/infinite-scroll/infinity.php
Comment (MattGeri) [inline modules/infinite-scroll/infinity.php:1833]: Is there a reason this is static? It's being called on a object in line 1807

File: modules/infinite-scroll/infinity.php
Comment (MattGeri) [inline modules/infinite-scroll/infinity.php:1702]: We should check if class_exists here as per previous instruction on some of the other issues we've worked on

File: modules/infinite-scroll/infinity.php
Comment (MattGeri) [inline modules/infinite-scroll/infinity.php:1572]: We should check if class_exists here as per previous instruction on some of the other issues we've worked on

File: modules/infinite-scroll/infinity.php
Comment (MattGeri) [inline modules/infinite-scroll/infinity.php:443]: We should check if class_exists here as per previous instruction on some of the other issues we've worked on

**Review by MattGeri — DISMISSED** [review]
Nice work so far Marco, added some initial feedback

File: modules/infinite-scroll/infinity.php
Comment (pereirinha) [reply modules/infinite-scroll/infinity.php]: @MattGeri 

That's a great catch.
For some reason, it made sense at some point. After the cleanup, I missed rethinking this one.

File: modules/infinite-scroll/infinity.php
Comment (pereirinha) [reply modules/infinite-scroll/infinity.php]: Addressed.

**Comment by pereirinha** [comment]
@MattGeri 

That's for the review.
I addressed your comments and this PR is ready for you to loop back.

File: modules/infinite-scroll/infinity.php
Comment (fgiannar) [inline modules/infinite-scroll/infinity.php:1779]: I’m a bit confused with the use of static  here as well as the null  check after setting it to null below.
Could you please elaborate on it?

**Review by fgiannar — CHANGES_REQUESTED** [review]
Hi there,

I tested this with **twentytwenty** theme and while in Standard Template mode everything works as expected, this is what I am seeing on **Transitional** mode:
https://d.pr/i/n5bAlM

Please also find one more comment inline.

Thanks!

File: modules/infinite-scroll/infinity.php
Comment (pereirinha) [reply modules/infinite-scroll/infinity.php:1779]: Hi @fgiannar 

Thanks for reviewing the code.

This method is supposed to be called twice in the runtime. The first time it runs happens when it's called from `amp_load_hooks()`. Because this is the first it runs, we declare a static `$template`. Then we check the contents for this variable, and if not set, we assign the contents of `amp_footer_template()`. In this method, we use output buffering to get a template for the footer. This first run warms the contents of the `$template`. 
The next time we run this method, it happens inside an output buffer callback. Because we declared the variable as being static, the `$template` now contains the footer markup so that the method will perform a slightly different action.
I tried a nested output buffer and have it done in one go, but I was having issues, likely because of the callback on the outer one.
This approach prevents having a property in this class whose only purpose is to be used by AMP workflow.

Please do let me know if I wasn't clear.

Comment (fgiannar) [reply modules/infinite-scroll/infinity.php:1779]: Hey @pereirinha 

Many thanks for the explanation, it's clear now :)
Just a heads up here: The `$template` static variable will keep its state inside different instances of our class. If this is the desired behaviour, please feel free to resolve this.

**Comment by pereirinha** [comment]
@fgiannar 

I tried to replicate the problem you shared, but I'm not able. Please check the following [screencast](https://drive.google.com/file/d/1BoUW6YurrS7w--CraGTiVHQCs9KvRa1Z/view).
Can it be the case that you have a plugin conflicting here?

File: modules/infinite-scroll/infinity.php
Comment (pereirinha) [reply modules/infinite-scroll/infinity.php:1779]: Yes, I'm aware of that, and it shouldn't be a problem.
I'm resolving. :)

**Comment by fgiannar** [comment]
@pereirinha Just a confirmation: In the screencast, when you refresh the page, you are already on page 2. Could you try Transitional mode and go to your homepage, so that next pages load via scrolling, and let me know if you could reproduce? 
Thanks for looking into this

**Comment by pereirinha** [comment]
@fgiannar 

That's a good point. I did try starting from the homepage, but I still don't see the issue.

**Comment by fgiannar** [comment]
@pereirinha hmm that's odd.. I tried in 3 different Jurassic Ninja fresh installations (with/without Gutenburg, with/without debugging etc) and could reproduce every time.
I am adding some screenshots of all the settings I am using:

**Jetpack settings:**
Screenshot 2020-06-22 at 1 04 45 PM

**Reading Settings**
Screenshot 2020-06-22 at 1 01 41 PM

**AMP settings**
Screenshot 2020-06-22 at 12 57 44 PM

Active Plugins: AMP, Jetpack Beta Tester
Error debugging: Off
Browser: Chrome latest version
Device: Mac Book Pro (2015)

It's also worth noting that while scrolling down, the url changes by displaying the current page and properly keeping the `amp` query param, but when I reach the near the bottom the `amp` param disappears (that's when the unformatted page section is observed) and re-appears when I reach the bottom.
Please check out this screencast regarding the above behaviour: https://d.pr/i/wiJmqa

**Comment by pereirinha** [comment]
Hi @fgiannar 

Sadly, I'm not getting the same experience as you're in the different environments I tested.

Thank you for sharing the screencast. I took the liberty to see in person the problem in your website, and I was able to realise that the issue happens because when we get to the end of the pages, the non-amp home page is loaded.

In my experience, what happens is that when you get to that last page, the [max-pages](https://amp.dev/documentation/components/amp-next-page/?format=websites#max-pages-(optional)) value kicks in and prevents loading the home page. Instead, it only shows a button that takes the visitor to the home page, if they choose to click it.

The `max-pages` attribute is set correctly in your system, but its ignored.

I pushed a change to this workflow. Instead of showing the home page button at the end, nothing will happen. You'll get to the end of the loop, and you'll see the footer. It's a less ideal situation, but it's perhaps safer. 

Please, let me know if it works.

Thank you.

**Comment by jeherve** [comment]
Internal reference: D45837-code

**Review by jeherve — CHANGES_REQUESTED** [review]
I rebased to address conflicts and issues with tests.

*********

I am running into the same issue that @fgiannar highlighted, also in Transitional mode. You can see it in action here:
https://jeherve.com/news/?amp

image

*********

In addition to this, the button title seems to have encoding issues (I'm using WP in French, where the button has accented characters).

*********

The page title also appears to be overwritten when you scroll down, and switches to "Older Posts"

*********

I cannot seem to access footer widgets anymore when using a theme like Twenty Twenty with footer widgets active. Outside of AMP views, we usually flip to a click behaviour when there are footer widgets on the site, so you can choose to trigger Infinite Scroll or view the widgets:
https://github.com/Automattic/jetpack/blob/8716e6f7c3aff8ae000c7a2ded9a0898f073a139/modules/theme-tools/compat/twentytwenty.php#L23

**Comment by pereirinha** [comment]
@jeherve @fgiannar 

Thanks for your inputs.

I've been trying to replicate the problem you have seen, but I've been failing to do it.

I did use the docker setup that comes with Jetpack to do the initial development of this PR. I then tried a different docker setup, [VVV](https://varyingvagrantvagrants.org/) and finally added this to my domain. In the other environments than Jetpack docker, I installed the Jetpack Beta and picked this branch.

[Here's a screencast](https://drive.google.com/file/d/1Nc5sKRv76-y_MJTckEWBKoC5x-3P3F4q/view?usp=sharing) where I show the settings and also what I get as the FE experience.

As for the French translation, I think that the issue is in the [translation itself](https://translate.wordpress.org/projects/wp-plugins/jetpack/dev/fr/default/?filters%5Bterm%5D=Older+posts&filters%5Bterm_scope%5D=scope_any&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Buser_login%5D=&filter=Apply+Filters&sort%5Bby%5D=priority&sort%5Bhow%5D=desc). I noticed similar cases in other strings.

Lastly, the AMP Infinite Scroll functionality isn't as flexible as the Jetpack option. The suggested approach relies on the [Recommendation box](https://amp.dev/documentation/components/amp-next-page/#recommendation-box-(more-suggested-links)) to show the Older posts button.

At this stage, I'm open for suggestion to be able to replicate the problem and from there access possible solutions.

**Comment by fgiannar** [comment]
Hi @pereirinha ,

I have tested this again and while I cannot reproduce the broken footer styles anymore, I can reproduce all issues @jeherve mentioned.

Some notes that might help:

In the latest stable branch, when I am visiting the amp version of the homepage, I see the following AMP related warning:
Screenshot 2020-07-20 at 12 41 37 PM

In the current branch there seem to be no warnings:
Screenshot 2020-07-20 at 12 44 11 PM

Visiting the **AMP Validation Error Index** page we can see various scripts removed:
Screenshot 2020-07-20 at 12 50 03 PM

My guess is that even we cannot reproduce the broken footer styles anymore, the reason behind it was that the AMP plugin was removing inline javascript, needed for the proper theme functionality.
Screenshot 2020-07-20 at 12 59 11 PM

By observing the removed scripts above, as well as in the screenshot provided by @jeherve we see that the AMP plugin is still removing Jetpack related inline scripts.

Testing Setup: https://github.com/Automattic/jetpack/pull/16048#issuecomment-647424499

I hope this helps :)

**Comment by pereirinha** [comment]
@fgiannar @jeherve 

Unfortunately, the depressed-pigeon is gone from the Jurassic ninja. On the bright side, we have now, and for the next 7 days,  a static donkey on https://static-donkey.jurassic.ninja/?amp.

In this environment, I saw that error in the console, which is introduced by a module called `notes` in a class called `Jetpack_Notifications` via `print_js` method. All this logic is unrelated to this work in this PR. I then disabled the infinite scroll to see if the problem persisted, and I couldn't see it anymore. Then I re-enabled the infinite scroll, to validate that the work in this PR was introducing it, but the error was gone for good. 

The Infinite Scroll seems to be working fine in this environment.

About the AMP Validation Error Index flagged by @fgiannar, can it be that they are in the were there since the attempt to use the latest stable branch? 

I updated the PR to fix the page title.

Let me know if you're experiencing any other problems.

**Review by fgiannar — DISMISSED** [review]
Hi @pereirinha ,

I tested again and can verify the only JS warning I am now getting is the one related with the `notes` module (`[multidoc-manager] - unallowed inline javascript`) and since you confirmed you have checked it out and is irrelevant with this PR, we are good to go!

Thanks for all your work!

**Comment by fgiannar** [comment]
@pereirinha ,

Actually, one final comment: Could you please add documentation for all the new filters implemented?
An example of adding docblocks to filters could be found [here](https://github.com/Automattic/jetpack/blob/a9dabaf0ffbbcf05b787b5496404cec50cb3fa7e/modules/infinite-scroll/infinity.php#L198-L207).
Once this is addressed, please feel free to merge!

Thank you!

**Comment by pereirinha** [comment]
@fgiannar 

Thank you for the turnaround, and I'm sorry for the delay. I took some days off.
I did add docblock information to the introduced filters and merged master.

I've tagged the `since` argument as `8.9.1`, as it makes sense to me, but I'm not sure if that's the intent. I'll defer to @jeherve on this.

**Comment by jeherve** [comment]
I've rebased to get rid of some unwanted commits.

> I've tagged the since argument as 8.9.1, as it makes sense to me, but I'm not sure if that's the intent.

I've updated those to 9.0.0 to match the next major release. 8.9.1 will only include bug fixes.

**Review by jeherve — COMMENTED** [review]
I ran into an issue on a site running Twenty Twenty, in AMP standard mode, and where Infinite Scroll was set to load with a button.

When clicking on the button to load the next batch of posts at the bottom of the page, page 2 loaded but this JavaScript error happened:

image

**Comment by jeherve** [comment]
Internal reference: D49477-code

**Review by jeherve — APPROVED** [review]
This seems to test well for me now. Merging.

**Comment by jeherve** [comment]
r213562-wpcom

**Comment by westonruter** [comment]
@jeherve This appears to not be working properly. AMP validation errors are being emitted on this AMP page, for example: https://alankent.me/2020/06/12/musings-about-5g-and-ar/amp/

The problem seems to be twofold:

1. The `<amp-next-page>` element is erroneously getting added when in legacy Reader mode.
2. The `amp-next-page` script is not getting injected because output buffer is being used to inject the tag and/or it's not getting detected during validation because WordPress.com is still on 1.4.x and in that old version the page is not sent through the post-processor. See https://github.com/ampproject/amp-wp/pull/3781.

**Comment by westonruter** [comment]
PR to fix: https://github.com/Automattic/jetpack/pull/17175

**Pull request merged**
